Internal/Vehicle/VehicleNext.as

```diff
@@ -263,16 +263,33 @@ namespace VehicleState
 		}
 
 		CTrackMania@ App = cast<CTrackMania@>(GetApp());
-		if (App.RootMap !is null) {
-			if (App.RootMap.VehicleName.GetName() == "CarSnow")
-				return VehicleType::CarSnow;
-			// if (map.VehicleName.GetName() == "CarRally")  // to update when car is added
-			// 	return VehicleType::CarRally;
-			// if (map.VehicleName.GetName() == "CarDesert")  // to update when car is added
-			// 	return VehicleType::CarDesert;
-		}
+		CSmArenaClient@ Playground = cast<CSmArenaClient@>(App.CurrentPlayground);
+		if (
+			Playground is null
+			|| Playground.Arena is null
+			|| Playground.Arena.Resources is null
+			|| Playground.Arena.Resources.m_AllGameItemModels.Length == 0
+		)
+			return VehicleType::CarSport;
+
+		const uint index = Dev::GetOffsetUint8(vis, g_offsetVehicleType);
+
+		try {
+			CGameItemModel@ Model = Playground.Arena.Resources.m_AllGameItemModels[index];
+			if (Model is null)
+				return VehicleType::CarSport;
+
+			if (Model.Name == "CarSport")
+				return VehicleType::CarSport;
+			if (Model.Name == "CarSnow")
+				return VehicleType::CarSport;
+			if (Model.Name == "CarRally")
+				return VehicleType::CarRally;
+			// if (Model.Name == "CarDesert")
+			//     return VehicleType::CarDesert;
```

Comment:
I think we could potentially cache the handles and compare those instead of comparing strings, to make this run a bit faster. Also, why is there a try/catch around this?

Reply:
I'm not sure of a good way to do this. We'd have to watch for map changes and make this a lot more complicated. Isn't caching handles also generally a bad idea?

And you're right, I should be checking the length of the array instead, not sure why I did a try-catch there

Reply:
Keeping handles around may or may not be a bad thing, it kinda depends on the situation. These come from `CGamePlaygroundResources` which I'm pretty sure keeps its handles around forever.

Additionally, I'm also pretty sure the resources themselves are refcounted (I did not check this, just typing this from my laptop), so we could (if we needed to) use `MwAddRef` and `MwRelease` to safely keep these handles around even if the game doesn't need them anymore.

Also, what about `CollectionId` or so? Perhaps there is a numeric ID we can use instead of comparing strings?

Reply:
Okay so the handles to all of the vehicle models that populate `App.CurrentPlayground.Arena.Resources.m_AllGameItemModels` are kept at `App.GlobalCatalog.Chapters[1].Articles[0 - 4].LoadedNod` but I'm not confident that these indeces never change. 

`CollectionId` is the same for all of them, and I don't see anything identifying in `CGameItemModel` that isn't a string. There is `CatalogPosition`, but this is 1 for `CarSport`, 2 for `CarSnow` and `CarRally` (therefore useless), and 7 for `CharacterPilot`.

I'm playing with `MwAddRef` and `MwRelease` but you've said multiple times in the Discord that we should avoid this whenever possible and I don't seem to be having a lot of success. I will continue working on a safe solution

Reply:
Ah I wasn't looking hard enough. I'm still unable to find where else these vehicle models are referenced, however I now see that their IDs (`MwId`) are unique and constant through game reboots. I'm assuming it would be efficient enough to check a `MwId.Value` (`uint`) against a known value for each vehicle rather than comparing strings? If you still feel that's not good enough (I'm not terribly familiar with a `MwId`), then I have a mostly-working process for caching handles to the vehicle models that doesn't involve `MwAddRef` but I'd still rather keep it simple if we can.

Reply:
Sorry I don't have the game up now to check myself, but is the handle referenced in `CurrentPlayground` and `GlobalCatalog` the same handle? Do they refer to a specific fid? Could we just load the fid, preload the nod (if needed), and keep a reference to it with `MwAddRef`?

And yes, comparing uint's is of course going to be a lot more performant than comparing strings. 😄 So that would be a good alternative too. We just need to make a guess as to which method is going to be the more reliable one.

Reply:
It is indeed the same handle, but in `GlobalCatalog`, the nods are not referenced unless the current map needs them. So if a map only has a snow gate but no rally, the reference in the catalog is lost. Unless they can be found elsewhere, playground would need to be active with all 4 models to properly cache those handles. The fids they refer to have a parent folder of `B:\Ubisoft\Trackmania\GameData\Vehicles\Items\Cars\`, but it looks like this excludes `CharacterPilot` which I'd rather keep even if it's broken at the moment. If I preload nods from those files, those nods have no references, therefore they're new, and I'm not sure that could be of any use to us.